### PR TITLE
Cm/update tree practice

### PR DESCRIPTION
**Pull request by CheezItMan**
If you don't mind looking this over.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,43 +1,16 @@
...
-2. Write a method that returns the count of non-operators (in this case, leaves) in the tree.
-    - _Added Fun: print the level of the tree that each leaf appears on._
-3. Write methods to print the tree in prefix and postfix notations.
-4. Write a method that returns whether or not a given operator exists in the tree.
-    - _Added fun: If the operator occurs more than once, return the count on how many times it occurs - you may return 0 if it doesn't exist in the tree._
+## Optionals
 
-## Code
-1. Code out solutions for problems 1-4 above.
+- `delete(value)` - This method removes a node from the tree.  Also add tests for the method.
```
Comment (kaidamasaki) [inline README.md:14]:
```suggestion
- `delete(value)` - This method removes a node from the tree.  Also add tests for this method.  
```

File: README.md [hunk 2]
```diff
@@ -1,43 +1,16 @@
...
-## Creating a binary manually tree with nodes
-1. Take a look at [`tree-practice.rb`](tree-practice.rb)
-2. Let's walk through how `3 + 2` was created as a binary tree.
-3. Let's walk through how `print_infix` works.
-4. Create `4 * 3 + 2 - 10 % 5` in code as a binary tree.
+- `add(value)` - This method adds a value to the Binary Search Tree
+- `find(value)` - This method returns true if the given value is in the tree and false otherwise.
+- `inorder` - This method returns an array of all the elements in the tree, in order.
+- `postorder` - This method returns an array of all the elements in a postorder fashion (left, right , root).
+- `preorder` - This method returns an array of all the elements in a preorder fashion (root, left, right).
```
Comment (kaidamasaki) [inline README.md:9]: Will they already know in-order, post-order and pre-order traversal?

Also, should those be `in_order`, `post_order` and `pre_order`?

File: README.md [hunk 3]
```diff
@@ -1,43 +1,16 @@
...
+- `bfs` - This method returns an array with the tree elements in a level-by-level order (Breadth-First traversal)
```
Comment (kaidamasaki) [inline README.md]:
```suggestion
- `bfs` - This method returns an array with the tree elements in a level-by-level order (Breadth-First Search)
```

File: lib/tree.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
+class TreeNode
```
Comment (kaidamasaki) [inline lib/tree.rb:1]: I tend to prefer making `left` and `right` be proper sub-trees.  It makes recursive solutions a little cleaner/more obvious.

If we do this we probably want to make the `left` and `right` methods protected.

File: test/test_helper.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+  
+require 'minitest'
+require 'minitest/autorun'
+require 'minitest/reporters'
+
+Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
+
+require_relative '../lib/tree'
```
Comment (kaidamasaki) [inline test/test_helper.rb]: Missing final newline.

```suggestion
require_relative '../lib/tree'

```

File: test/tree_test.rb [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      tree.add(2)
+      expect(tree.height).must_equal 3
+
+      tree.add(3)
+      expect(tree.height).must_equal 3
+
+      tree.add(4)
+      expect(tree.height).must_equal 4
+    end
+end
```
Comment (kaidamasaki) [inline test/tree_test.rb:83]: Missing final newline.

```suggestion
end

```

File: lib/tree.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+  # Time Complexity: 
+  # Space Complexity: 
+  def find(val)
+    raise NotImplementedError
+  end
+
+  # Time Complexity: 
+  # Space Complexity: 
+  def inorder
```
Comment (kaidamasaki) [inline lib/tree.rb:33]: If you change the names above remember to update here!

File: test/tree_test.rb [hunk 8]
```diff
@@ -0,0 +1,104 @@
+require_relative 'test_helper'
+
+
+Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
+
+describe Tree do
+  let (:tree) {Tree.new}
+
+  def add_nodes(tree)
```
Comment (kaidamasaki) [inline test/tree_test.rb]: I think I would prefer a `let (:tree_with_nodes)`...

File: README.md (hunk 2 above)
Comment (CheezItMan) [reply README.md:9]: Yes, and I've always seen inorder, preorder and postorder written as whole words.

**Pull request merged**